File: 347-top-k-frequent-elements/347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> hm = new HashMap<>();
        for(int num : nums){
            int count = hm.getOrDefault(num, 0) + 1;
            hm.put(num, count);
        }
        
        int[] res = new int[k];
        int index = 0;
        for(int i = 0; i < k; i++){
            int max = -1;
            int maxKey = 0;
            for(int key : hm.keySet()){
                if(hm.get(key) > max){
                    max = hm.get(key);
                    maxKey = key;
                }
            }
            res[index] = maxKey;
            hm.put(maxKey, -2);
            index++;
        }
        return res;
    }
}

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums, k):
        #make buckets of each frequency, an element can appear at most k times
        bucket = [[] for _ in range(len(nums) + 1)]
        #make a counter of the items
        Count = Counter(nums)  
        for num, freq in Count.items():
            bucket[freq].append(num) 
        
        flat_list = []
        for b in bucket:
            for item in b:
                flat_list.append(item)
        return flat_list[len(flat_list)-k:]